src/plugins/discover/public/application/main/state_management/discover_data_state_container.ts

```diff
@@ -293,6 +293,13 @@ export function getDataStateContainer({
               ...commonFetchDeps,
             },
             async () => {
+              const { resetDefaultProfileState: currentResetDefaultProfileState } =
+                internalStateContainer.getState();
+
+              if (currentResetDefaultProfileState.resetId !== resetDefaultProfileState.resetId) {
+                return;
+              }
```

Comment:
The issue was that there is a 100ms delay between when we update the app state data source to the new data view in `changeDataView` (and trigger default state reset) and when the `fetch$` callback is actually called (due to the debounce in `getFetch$`). If within that 100ms period, the previous fetch finishes, this code was being run and clearing out the default state reset here: https://github.com/elastic/kibana/blob/2213f4319d9445b597bba9f726747348619b4aa8/src/plugins/discover/public/application/main/state_management/discover_data_state_container.ts#L314-L320

This causes the following code which checks for the default state reset above to receive all `false` values, and the correct default state does not get set: https://github.com/elastic/kibana/blob/2213f4319d9445b597bba9f726747348619b4aa8/src/plugins/discover/public/application/main/state_management/discover_data_state_container.ts#L268-L271

The solution was to ensure the default state reset call associated with this fetch is actually the latest one when this code is run, by using a `resetId` that gets updated on each reset call. This is not the most elegant solution to this issue and is a symptom of a deeper issue, but it was the least invasive one I could come up with. I believe the long term / correct solution to this issue is to refactor Discover data fetching so logic runs sequentially and is easier to reason about.